### PR TITLE
Add experimental support to CronJob

### PR DESCRIPTION
**Pull request by drgarcia1986**

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -180,6 +180,25 @@ automatic migrations by adding this line to the `Procfile`:
 Note that a failing release will prevent the rolling update from happening, so
 you have to keep compatibility with old code.
 
+### CronJob
+
+**Q: How to create a CronJob?**
+
+    $ teresa app create <app-name> --team <team-name> --process-type cron
+
+Make sure to have the `cron` key with the cronjob command.
```
Comment (yagonobre) [inline README.md]: Maybe
```
Make sure to have the `cron` entry in procfile with the cronjob command.
```

Comment (drgarcia1986) [reply README.md]: I followed the previous example [114] _You also have to adjust the `Procfile` to have a corresponding `worker` key._ , but, lgtm to change

File: pkg/server/app/app.go [hunk 2]
```diff
@@ -119,8 +119,10 @@ func (ops *AppOperations) Create(user *database.User, app *App) (Err error) {
 		return teresa_errors.NewInternalServerError(err)
 	}
 
-	if err := ops.kops.CreateOrUpdateAutoscale(app); err != nil {
-		return teresa_errors.New(ErrInvalidAutoscale, err)
+	if app.ProcessType != ProcessTypeCron {
```
[the remark this answers is not in the document]
Comment (yagonobre) [reply pkg/server/app/app.go]: but decrease one level of indentation

File: pkg/server/deploy/deploy.go [hunk 3]
```diff
@@ -82,26 +84,11 @@ func (ops *DeployOperations) Deploy(ctx context.Context, user *database.User, ap
...
 		slugURL := fmt.Sprintf("%s/slug.tgz", buildDest)
-		releaseCmd := confFiles.Procfile[ProcfileReleaseCmd]
-		if confFiles.Procfile != nil && releaseCmd != "" {
-			if err := ops.runReleaseCmd(a, deployId, slugURL, w); err != nil {
-				errChan <- err
-				log.WithError(err).WithField("id", deployId).Errorf("Running release command %s in app %s", releaseCmd, appName)
-				return
-			}
+		if a.ProcessType == app.ProcessTypeCron {
+			ops.createOrUpdateCronJob(a, confFiles, w, errChan, slugURL, description)
```
[the remark this answers is not in the document]
Comment (drgarcia1986) [reply pkg/server/deploy/deploy.go]: I'm not sure, that method calls `CreateOrUpdateCronJob` from `K8sOperations` interface, effectively its create or update the cronjob object on kubernetes, depends on if the cronjob objects already exist or not.

File: pkg/server/deploy/deploy.go [hunk 4]
```diff
@@ -139,7 +126,50 @@ func (ops *DeployOperations) buildLimits() *spec.ContainerLimits {
 	}
 }
 
-func (ops *DeployOperations) createDeploy(a *app.App, tYaml *spec.TeresaYaml, description, slugURL string) error {
+func (ops *DeployOperations) effectivelyDeploy(a *app.App, confFiles *DeployConfigFiles, w io.Writer, errChan chan error, slugURL, deployId, description string) {
```
[the remark this answers is not in the document]
Comment (drgarcia1986) [reply pkg/server/deploy/deploy.go]: `effectivelyDeploy` is a terrible name, I made two changes, first ~effectivelyDeploy~ -> `createOrUpdateDeploy`, second I removed the method `createOrUpdateK8sDeploy` and move the logic to `createOrUpdateDeploy` method. What do you think?

File: pkg/server/app/app.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (drgarcia1986) [reply pkg/server/app/app.go]: I didn't say that I'm not going to do 😂.

File: README.md (hunk 1 above)
Comment (yagonobre) [reply README.md]: yes, keep the pattern sounds better

File: pkg/server/deploy/deploy.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (yagonobre) [reply pkg/server/deploy/deploy.go]: Great! 👏👏

File: pkg/server/deploy/deploy.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (yagonobre) [reply pkg/server/deploy/deploy.go]: Yes, my point is that the `ops.CreateOrUpdateCronJob` always creates a new cronjob spec.

[the remark this answers is not in the document]
Comment (drgarcia1986) [reply pkg/server/deploy/deploy.go]: nops, its call `k8s.CreateOrUpdateCronJob` on line 182

File: pkg/server/app/app_test.go [hunk 5]
```diff
@@ -269,6 +271,28 @@ func TestAppOperationsCreate(t *testing.T) {
...
+		Name:  name,
+		Users: []database.User{*user},
+	}
+
+	if err := ops.Create(user, app); err != nil {
+		t.Fatal("error creating app: ", err)
+	}
+
+	if fakeK8s.CreateOrUpdateAutoscaleWasCalled {
+		t.Error("expected not hpa for crons, but was created")
```
Comment (aguerra) [inline pkg/server/app/app_test.go]: expected no hpa

File: pkg/server/spec/deploy.go [hunk 6]
```diff
@@ -84,11 +89,7 @@ func NewDeploy(imgs *SlugImages, description, slugURL string, rhl int, a *app.Ap
 	}
 
 	if tYaml != nil {
-		ds.TeresaYaml = TeresaYaml{
-			HealthCheck:   tYaml.HealthCheck,
-			RollingUpdate: tYaml.RollingUpdate,
-			Lifecycle:     tYaml.Lifecycle,
-		}
+		ds.TeresaYaml = *tYaml
```
Comment (aguerra) [inline pkg/server/spec/deploy.go]: :clap:

File: README.md [hunk 7]
```diff
@@ -180,6 +180,25 @@ automatic migrations by adding this line to the `Procfile`:
...
+**Q: How to define the schedule of the CronJob?**
+
+You can define the schedule of cron by adding this lines to `teresa.yaml`:
+
+```
+cron:
+  schedule: "*/30 * * * *"
+```
+
+You can change the schedule on cron on every deploy.
```
Comment (aguerra) [inline README.md]: cron schedule sounds better to me

File: pkg/server/spec/deploy.go [hunk 8]
```diff
@@ -40,10 +40,15 @@ type Lifecycle struct {
 	PreStop *PreStop `yaml:"preStop,omitempty"`
 }
 
+type CronArgs struct {
```
[the remark this answers is not in the document]
Comment (drgarcia1986) [reply pkg/server/spec/deploy.go]: Remove struct and use a simple string? 
For code looks good, but, for yaml I think its better to create another level, f.ex:
```yaml
# teresa.yaml
cron:
    schedule: "...."
```
What do you think?

[the remark this answers is not in the document]
Comment (yagonobre) [reply pkg/server/spec/deploy.go]: Nice!

File: pkg/server/deploy/deploy.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (yagonobre) [reply pkg/server/deploy/deploy.go]: I'm talking about the spec not the CronJob in k8s

**Pull request merged**